### PR TITLE
Update `parser-standard` to compiler's parser

### PR DESCRIPTION
**Pull request by freemagma**
Currently, `parser-standard` is quite out of date with jane street's compiler's parser. This PR gets them in close to in line as possible, and adds some scripts to make future updates easier.

Also disables the windows CI, because of a strange and difficult to debug jane syntax error.

Notes for reviewer:
- `parser-jane` is just a copy of the `flambda-backend` parser with modes and modalities moved out of jane syntax (changes which will eventually be merged but are currently in progress)
- `parser-standard` has changed to be very close to an exact copy of `parser-jane`, with some patches applied
- The only other file that has changed much is `Normalize_std_ast.mli`, which now has to account for the up-to-date way that jane syntax encodes attributes. This file will become a lot simpler if jane syntax is fully removed in the future

### CODE REVIEW

File: .github/workflows/build-others.yml [hunk 1]
```diff
@@ -18,7 +18,6 @@ jobs:
       matrix:
         os:
           - macos-latest
-          - windows-latest
```
Comment (goldfirere) [inline .github/workflows/build-others.yml:21]: Maybe leave a comment here explaining that we'd like windows support back?

File: vendor/parser-jane/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# parser-jane
+This directory contains a direct copy of files from Jane Street's compiler's
```
Comment (goldfirere) [inline vendor/parser-jane/README.md:2]: It would probably good to preserve the git hash of the base commit in flambda-backend. I see us wanting to be able to know what compiler ocamlformat is based offf of.

File: vendor/parser-jane/lexer.mll
Comment (goldfirere) [inline vendor/parser-jane/lexer.mll:1]: This looks different from the compiler's lexer. Is that intentional?

File: vendor/parser-jane/update.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+if [[ "$#" == "1" ]] ; then
+    flambda_backend_dir="$1"
+else
+    echo "Wrong number of arguments"
+    exit 1
+fi
+
+parsing_dir="${flambda_backend_dir}/ocaml/parsing"
+utils_dir="${flambda_backend_dir}/ocaml/utils"
+lex_dir="${flambda_backend_dir}/ocaml/lex"
```
Comment (goldfirere) [inline vendor/parser-jane/update.sh:13]: I think this is the problem with the lexer. The correct lexer is in `parsing`. The lexer in `lex` is actually the lexer for .mll files, not .ml files. :)

File: vendor/parser-jane/repatch.sh
Comment (goldfirere) [inline vendor/parser-jane/repatch.sh:1]: Should this also update the relevant files in ocaml-common?

File: lib/Normalize_std_ast.ml [hunk 4]
```diff
@@ -12,116 +12,55 @@
...
 
+let is_builtin_jane_syntax attr =
+  let name = attr.attr_name.txt in
+  String.is_prefix ~prefix:"jane.erasable._builtin" name
+
 let is_erasable_jane_syntax attr =
-  String.is_prefix ~prefix:"jane.erasable." attr.attr_name.txt
+  let name = attr.attr_name.txt in
+  String.is_prefix ~prefix:"jane.erasable." name
+  || String.equal "extension.curry" name
```
Comment (goldfirere) [inline lib/Normalize_std_ast.ml:30]: Why is this here? It seems new in this patch.

File: lib/Normalize_std_ast.ml [hunk 5]
```diff
@@ -12,116 +12,55 @@
...
+
+let is_erasable_builtin_jane_syntax attr =
```
Comment (goldfirere) [inline lib/Normalize_std_ast.ml]: This looks identical to `is_builtin_jane_syntax`.

File: lib/Normalize_std_ast.ml [hunk 6]
```diff
@@ -209,27 +148,44 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
...
-  let attributes (m : Ast_mapper.mapper) (atrs : attribute list) =
+  let map_attributes_no_sort (m : Ast_mapper.mapper) (atrs : attribute list)
+      =
     let atrs =
       if erase_jane_syntax then
         List.filter atrs ~f:(fun a -> not (is_erasable_jane_syntax a))
-      else atrs
+      else
+        List.filter atrs ~f:(fun a ->
+            not (is_erasable_builtin_jane_syntax a) )
```
Comment (goldfirere) [inline lib/Normalize_std_ast.ml]: This is about n-ary arrows, right? Please comment this somewhere, maybe in `is_erasable_builtin_jane_syntax`. And the whole thing isn't really about erasability, but rather the fact that ocamlformat really does something sketchy around n-ary arrows. So I would be more direct about the name: `is_n_ary_arrows_attribute` or something. Because I think if we had other `jane.erasable.builtin_` things, they wouldn't get the same treatment here.

File: lib/Normalize_std_ast.ml [hunk 7]
```diff
@@ -209,27 +148,44 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
...
-    | Pexp_poly ({pexp_desc= Pexp_constraint (e, t); _}, None) ->
+    | Pexp_apply
+        ( {pexp_desc= Pexp_extension ({txt= "extension.exclave"; _}, _); _}
+        , [(Nolabel, expr)] )
+      when erase_jane_syntax ->
+        m.expr m expr
+    | Pexp_poly ({pexp_desc= Pexp_constraint (e, Some t, []); _}, None) ->
         m.expr m {exp with pexp_desc= Pexp_poly (e, Some t)}
-    | Pexp_constraint (e, {ptyp_desc= Ptyp_poly ([], _t); _}) -> m.expr m e
+    | Pexp_constraint (exp1, None, _ :: _) when erase_jane_syntax ->
```
Comment (goldfirere) [inline lib/Normalize_std_ast.ml:188]:
```suggestion
    (* mode coercion *)
    | Pexp_constraint (exp1, None, _ :: _) when erase_jane_syntax ->
```

File: lib/Normalize_std_ast.ml [hunk 8]
```diff
@@ -287,20 +249,23 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
...
-    let typ = {typ with ptyp_loc_stack= []} in
     let typ =
       { typ with
-        ptyp_attributes=
-          convert_legacy_jane_street_local_annotations ~segment:Type
-            typ.ptyp_attributes }
+        ptyp_loc_stack= []
+      ; ptyp_attributes=
+          ( typ.ptyp_attributes
+          |> if erase_jane_syntax then map_attributes_no_sort m else Fn.id )
```
Comment (goldfirere) [inline lib/Normalize_std_ast.ml:268]: This seems like it might treat doc comments wrongly, because the decision of whether to strip them is dependent on `erase_jane_syntax`. We probably don't want to strip doc comments here, but `map_attributes_no_sort` appears to.

File: lib/Normalize_std_ast.ml [hunk 9]
```diff
@@ -209,27 +148,44 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
...
+  in
+  let attributes (m : Ast_mapper.mapper) (atrs : attribute list) =
+    sort_attributes (map_attributes_no_sort m atrs)
   in
   let expr (m : Ast_mapper.mapper) exp =
-    let exp = {exp with pexp_loc_stack= []} in
+    let exp =
+      { exp with
+        pexp_loc_stack= []
+      ; pexp_attributes= map_attributes_no_sort m exp.pexp_attributes }
```
Comment (goldfirere) [inline lib/Normalize_std_ast.ml]: After we're done killing off jane-syntax, this can be removed. Maybe leave a comment? I would start it `(* CR jane-syntax: ... *)`. This is different from the convention we use internally in two ways:

1. We actually commit code with `CR`. So `CR` really means something like `CR-someday`.
2. We label with the project, not the person. This makes these items easier to find later.

Feel free to add more such comments if there are other items to return to.

Comment (goldfirere) [inline lib/Normalize_std_ast.ml]: This step will conditionally strip doc comments, no? We don't necessarily want to do that here; the previous code didn't.

File: lib/Normalize_std_ast.ml [hunk 10]
```diff
@@ -369,46 +336,49 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
...
-    in
-    Ast_mapper.default_mapper.constructor_declaration m
-      {decl with pcd_args= args}
-  in
   let type_declaration (m : Ast_mapper.mapper) decl =
     let ptype_attributes =
       decl.ptype_attributes
-      |> List.map ~f:normalize_immediate_annot_and_attrs
+      |> List.filter_map ~f:normalize_immediate_annot_and_attrs
+      |> if erase_jane_syntax then map_attributes_no_sort m else Fn.id
```
Comment (goldfirere) [inline lib/Normalize_std_ast.ml:346]: This is another one that should be removed once jane-syntax is gone

File: lib/Normalize_std_ast.ml [hunk 11]
```diff
@@ -369,46 +336,49 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
...
   in
+  let modes (m : Ast_mapper.mapper) ms =
+    Ast_mapper.default_mapper.modes m (if erase_jane_syntax then [] else ms)
+  in
+  let modalities (m : Ast_mapper.mapper) ms =
+    Ast_mapper.default_mapper.modalities m
+      (if erase_jane_syntax then [] else ms)
+  in
+  let value_binding (m : Ast_mapper.mapper) vb =
+    let vb =
```
Comment (goldfirere) [inline lib/Normalize_std_ast.ml:358]: This doesn't quite look right. It looks like it's moving modes from the expression to the value binding, unconditionally (there's no `erase_jane_syntax` check). Why do this? And is there a missing test case?

File: lib/Normalize_std_ast.ml [hunk 12]
```diff
@@ -369,46 +336,49 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
...
+                {vb.pvb_expr with pexp_desc= Pexp_constraint (exp, cty, [])}
+          in
+          {vb with pvb_modes= vb.pvb_modes @ modes; pvb_expr}
+      | _ -> vb
+    in
+    Ast_mapper.default_mapper.value_binding m vb
+  in
+  let constructor_declaration (m : Ast_mapper.mapper) cd =
+    ( if erase_jane_syntax then
+        {cd with pcd_attributes= map_attributes_no_sort m cd.pcd_attributes}
```
Comment (goldfirere) [inline lib/Normalize_std_ast.ml:378]: These steps can be stopped after jane-syntax is dead -- and probably do the wrong thing with doc comments.

File: vendor/parser-jane/update.sh (hunk 3 above)
Comment (freemagma) [reply vendor/parser-jane/update.sh:13]: ahh i see. fixed.

File: vendor/parser-jane/repatch.sh
Comment (freemagma) [reply vendor/parser-jane/repatch.sh:1]: probably, yes. done

File: lib/Normalize_std_ast.ml (hunk 4 above)
Comment (freemagma) [reply lib/Normalize_std_ast.ml:30]: added a CR to explain. basically, ocamlformat formats `int -> (int -> int)` to `int -> int -> int` when removing jane syntax, which causes the removal of a `extension.curry` attribute. so we need to ignore this attribute when erasing jane syntax

File: lib/Normalize_std_ast.ml (hunk 6 above)
Comment (freemagma) [reply lib/Normalize_std_ast.ml]: Refactored so this isn't necessary here anymore

File: lib/Normalize_std_ast.ml (hunk 7 above)
Comment (freemagma) [reply lib/Normalize_std_ast.ml:188]: Added a more descriptive comment than your suggestion

File: lib/Normalize_std_ast.ml (hunk 8 above)
Comment (freemagma) [reply lib/Normalize_std_ast.ml:268]: Not sure I agree with any of your points about doc comments. The mapper removes doc comments when `ignore_doc_comments` is true, the same as it always did. The difference here is that if `erase_jane_syntax` is true, it will do that step right now -- otherwise, it will still do it in the `Ast_mapper.default_mapper.typ` call below (since that call will invoke our `attributes` mapper on `ptyp_attributes`). The reason we need a special case is because `Ast_mapper.default_mapper.typ` hides the jane-syntax related attributes away before we can get to them with the `attributes` mapper.

File: lib/Normalize_std_ast.ml (hunk 9 above)
Comment (freemagma) [reply lib/Normalize_std_ast.ml]: refactored this a bit, but added comments everywhere necessary

File: lib/Normalize_std_ast.ml (hunk 11 above)
Comment (freemagma) [reply lib/Normalize_std_ast.ml:358]: Added a comment to explain. Basically we currently format `let x = local_ ("" : string)` into `let local_ x = ("" : string)`, so this is mapping is needed

File: lib/Normalize_std_ast.ml (hunk 8 above)
Comment (goldfirere) [reply lib/Normalize_std_ast.ml:268]: We spoke about this. I'm satisfied.
